holopy/scattering/scatterer/composite.py

```diff
@@ -74,45 +85,71 @@ def get_component_list(self):
-            for key, par in scatterer.parameters.items():
-                for index, par_check in enumerate(pars + [None]):
+    def update_ties(self):
+        for fullkey, par in self.raw_parameters.items():
+            if fullkey not in sum(self.ties.values(), []):
+                # not already in the list of ties, so check if it should be
+                reference_parameters = dict_without(self.parameters, fullkey)
+                for ref_key, ref_par in reference_parameters.items():
                     # can't simply check par in parameters because then two
                     # priors defined separately, but identically will match
                     # whereas this way they are counted as separate objects.
-                    if par is par_check and not isinstance(par, Number):
-                        # par is a prior and already exists in pars
+                    if par is ref_par and not isinstance(par, Number):
+                        if ref_key in self.ties.keys():
+                            self.ties[ref_key].append(fullkey)
+                        else:
+                            subkey = fullkey.split(':', 1)[1]
+                            if subkey not in self.ties.keys():
+                                self.ties[subkey] = [ref_key, fullkey]
+                            else:
+                                self.ties[fullkey] = [ref_key, fullkey]
                         break
-                if par_check is None:
-                    # loop finished - par is not tied.
-                    names.append('{0}:{1}'.format(i,key))
-                    pars.append(par)
-                else:
-                    # loop encountered a break - parameter is tied
-                    names[index] = key
-        return {key:val for key, val in zip(names, pars)}
 
+    @property
+    def raw_parameters(self):
+        parameters = {}
+        for i, scatterer in enumerate(self.scatterers):
+            single_scatterer_parameters = {'{0}:{1}'.format(i, key): val
+                            for key, val in scatterer.parameters.items()}
+            parameters.update(single_scatterer_parameters)
+        return parameters
+
+    @property
+    def parameters(self):
+        parameters = self.raw_parameters
+        for tied_name, raw_names in self.ties.items():
+            for raw_name in raw_names:
+                if raw_name not in parameters.keys():
+                    msg = 'Tied parameter {} not present in raw \
+                           parameters {}.'.format(raw_name, parameters.keys())
+                    raise ValueError(msg)
+            tied_val = parameters[raw_names[0]]
+            for raw_name in raw_names:
+                if not parameters[raw_name] == tied_val:
+                    msg = 'Tied parameters {} and {} are not equal.'.format(
+                            parameters[raw_name], tied_val)
+                    raise ValueError(msg)
+                del parameters[raw_name]
```

Comment:
OK. I'm a little confused what this function is doing. It seems that you are 
1. Copying ```raw_parameters```, which contains duplicates.
2. Removing a parameter one time when it appears in ```self.ties```.
I'm confused as to why this works, as it seems to me that this assumes that tied parameters only appear in ```raw_parameters``` two times at most (e.g. I think this will fail if there are 3 spheres all tied to share the same index). If that's the case, why doesn't something like this work?
```python
@property
def parameters(self):
    ties_reversed = {raw_name: tied_name for tied_name, raw_name in self.ties.items()}
    parameters = {
        name: parameter for name, parameter in self.raw_parameters.items()
        if name not in ties_reversed else ties_reversed[name]: parameter}
    return parameters
```

Or am I missing something?